penumbra-zone/web
# Reject `Swap` approvals with foreign claim addresses

From #548:

> The claim address is specified in the SwapPlan. It should always be an address the user controls.
>
> **The extension's approval code should check that any submitted SwapPlan has a claim address controlled by the user, and automatically reject it if not.**
>
> Adding this check means that we don't need to display the claim address to the user in the approval view.

## In this PR
- In the `CustodyService`'s `authorize` implementation, assert that transaction swaps (if there are any) contain claim addresses controlled by the current user.
  - To keep things clean, and because we may want to add more assertions as we think of other cases like this, I put the assertion code and its tests in a separate file from the `authorize` implementation.
- Rename the poorly named `isControlledAddress` helper to `getAddressIndexByAddress` to better represent what it does.
- Create a new `isControlledAddress` helper that actually returns a boolean. 
- Remove the claim address from `<SwapViewComponent />`.

## Manual testing

The code:

```diff
diff --git a/apps/minifront/src/state/swap.ts b/apps/minifront/src/state/swap.ts
index 49ca99fe..52a867f4 100644
--- a/apps/minifront/src/state/swap.ts
+++ b/apps/minifront/src/state/swap.ts
@@ -24,6 +24,7 @@ import { getAssetId } from '@penumbra-zone/getters/src/metadata';
 import { getSwapCommitmentFromTx } from '@penumbra-zone/getters/src/transaction';
 import { getAddressIndex } from '@penumbra-zone/getters/src/address-view';
 import { toBaseUnit } from '@penumbra-zone/types/src/lo-hi';
+import { bech32ToAddress } from '@penumbra-zone/bech32/src/address';

 export interface SimulateSwapResult {
   output: ValueView;
@@ -163,7 +164,13 @@ const assembleSwapRequest = async ({ assetIn, amount, assetOut }: SwapSlice) =>
           ),
           assetId: getAssetIdFromValueView(assetIn.balanceView),
         },
-        claimAddress: await getAddressByIndex(addressIndex.account),
+        // claimAddress: await getAddressByIndex(addressIndex.account),
+        claimAddress: {
+          inner: bech32ToAddress(
+            'penumbra1tj9s0dh8ymphnw2qgrva57wpcmpr38e9a8w05t5c8dyynr7dt58hgnare7tunqyuryc7yag084rqxrsrj9yx8uk9yhs8lmg64hl993wn3rphlvan3paxdd6suyhwnpkd6cwzes',
+          ),
+        },
+
         // TODO: Calculate this properly in subsequent PR
         //       Asset Id should almost certainly be upenumbra,
         //       may need to indicate native denom in registry
```

The result:


https://github.com/penumbra-zone/web/assets/1121544/1a4caf95-9a73-4f11-a11f-2ee58be57574


Note that it just says "Transaction canceled," rather than showing the error message about the website stealing your funds. That's because it throws a `ConnectError` with `Code.PermissionDenied`, which minifront catches and displays as "Transaction canceled." 

Closes #548
